db/fetcher/fetcher_acp.go

```diff
@@ -0,0 +1,39 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package fetcher
+
+import (
+	"context"
+
+	"github.com/sourcenetwork/defradb/acp"
+	"github.com/sourcenetwork/defradb/db/permission"
+)
+
+// runDocReadPermissionCheck handles the checking (while fetching) if the document has read access
+// or not, according to our access logic based on weather (1) the request is permissioned,
+// (2) the collection is permissioned (has a policy), (3) acp module exists.
+func (df *DocumentFetcher) runDocReadPermissionCheck(ctx context.Context) error {
+	hasPermission, err := permission.CheckDocPermissionedAccessOnCollection(
```

Comment:
question: How expensive do you see this call being?  Do you ever see a need for it to be aysnc at somepoint?

suggestion: It does at first glance look like it might be better to instead do the acp stuff in a new node, the fetcher is already quite painful to maintain, and a new node feels more aligned with the spirit of the planner package. Less importantly, this also allows turning it on/off or replacing it to be evaluated during plan construction, instead of the main iteration loop (slightly easier for us, and the machine).

Reply:
Yes the asyc aspects are on the roadmap. For now this uses the embedding local acp module.

Reply:
Do you see the location code flow around the calling of this function to remain the same when that time comes? Or is the code flow here likely to be very temporary?  And do you have a rough idea of what that new code structure might look like? 

The thought of a blocking async call per document inside the fetcher is not a happy one.

Reply:
replying here: https://github.com/sourcenetwork/defradb/pull/2338